x/leverage/keeper/keeper.go

```diff
@@ -271,3 +265,181 @@ func (k Keeper) GetCollateralSetting(ctx sdk.Context, borrowerAddr sdk.AccAddres
+// for a selected denomination of uToken collateral. If the borrower is not over their borrow limit, or
+// the repayment or reward denominations are invalid, an error is returned. If the attempted repayment
+// is greater than the amount owed or the maximum that can be repaid due to parameters (close factor)
+// then a partial liquidation, equal to the maximum valid amount, is performed. The same occurs if the
+// value of collateral in the selected reward denomination cannot cover the proposed repayment.
+// Because partial liquidation is possible and exchange rates vary, LiquidateBorrow returns the actual
+// amount of tokens repaid and uTokens rewarded (in that order).
+func (k Keeper) LiquidateBorrow(
+	ctx sdk.Context, liquidatorAddr, borrowerAddr sdk.AccAddress, repayment sdk.Coin, rewardDenom string,
+) (sdk.Int, sdk.Int, error) {
+	if !repayment.IsValid() {
+		return sdk.ZeroInt(), sdk.ZeroInt(), sdkerrors.Wrap(types.ErrInvalidAsset, repayment.String())
+	}
+	if !k.IsAcceptedUToken(ctx, rewardDenom) {
+		return sdk.ZeroInt(), sdk.ZeroInt(), sdkerrors.Wrap(types.ErrInvalidAsset, rewardDenom)
+	}
+
+	// Get total borrowed by borrower (all denoms)
+	borrowed, err := k.GetBorrowerBorrows(ctx, borrowerAddr)
+	if err != nil {
+		return sdk.ZeroInt(), sdk.ZeroInt(), err
+	}
+
+	// Get borrower uToken balances, for all uToken denoms enabled as collateral
+	collateral := k.GetBorrowerCollateral(ctx, borrowerAddr)
+	if err != nil {
+		return sdk.ZeroInt(), sdk.ZeroInt(), err
+	}
+
+	// Use oracle helper functions to find total borrowed value in USD
+	borrowValue, err := k.TotalPrice(ctx, borrowed)
+	if err != nil {
+		return sdk.ZeroInt(), sdk.ZeroInt(), err
+	}
+
+	// Use collateral weights to compute borrow limit from enabled collateral
+	borrowLimit, err := k.CalculateBorrowLimit(ctx, collateral)
+	if err != nil {
+		return sdk.ZeroInt(), sdk.ZeroInt(), err
+	}
+
+	// Confirm borrower's eligibility for liquidation
+	if borrowLimit.GTE(borrowValue) {
+		return sdk.ZeroInt(), sdk.ZeroInt(), sdkerrors.Wrap(types.ErrLiquidationIneligible, borrowerAddr.String())
+	}
+
+	// Get reward-specific incentive and dynamic close factor
+	baseRewardDenom := k.FromUTokenToTokenDenom(ctx, rewardDenom)
+	liquidationIncentive, closeFactor, err := k.LiquidationParams(ctx, baseRewardDenom, borrowValue, borrowLimit)
+	if err != nil {
+		return sdk.ZeroInt(), sdk.ZeroInt(), err
+	}
+
+	// Get liquidator's available balance of base asset to repay
+	liquidatorBalance := k.bankKeeper.GetBalance(ctx, liquidatorAddr, repayment.Denom)
+	// Return specific error on zero liquidator balance
+	if liquidatorBalance.Amount.IsZero() {
+		return sdk.ZeroInt(), sdk.ZeroInt(), sdkerrors.Wrap(types.ErrLiquidatorBalanceZero, liquidatorBalance.String())
+	}
+	// Repayment cannot exceed liquidator's available balance
+	repayment.Amount = sdk.MinInt(repayment.Amount, liquidatorBalance.Amount)
```

Comment:
wdyt about not modifying the input param `repayment` and just keeping track internally of `repaymentAmt` as an `sdk.Int`?

Reply:
That would work, the new `repaymentAmt` would take the place of `repayment.Amount` everywhere, e.g.

```go
// was repayValue, err := k.Price(ctx, repayment)
repayValue, err := k.Price(ctx, sdk.NewCoin(repaymentAmt, repayment.Denom))
```

Is this just for clarity?

I double-checked that the input `sdk.Coin` isn't acting like a pointer, so outside functions passing in a coin don't have to worry about the amount being modified. That would have been a problem in `msg_server.go`, where the both original (attempted) and final (successful) repayment amounts are used for logging.

Reply:
Yeah I just feel uneasy about modifying input parameters. I would attempt to use a local variable instead, either `Int` or `Coin`.